src/main/kotlin/org/misarch/shipment/service/ShipmentService.kt

```diff
@@ -170,21 +180,46 @@ class ShipmentService(
         })
         val address = addressRepository.findById(input.addressId).awaitSingle()
         val webClient = WebClient.create()
-        webClient.post().uri(shipmentProviderConfigurationProperties.endpoint).bodyValue(
-            ShipmentProviderShipmentDefinition(
-                shipmentId = savedShipment.id!!,
-                ref = shipmentMethod.externalReference,
-                quantity = quantity,
-                weight = weight,
-                address = AddressDefinition(
-                    street1 = address.street1,
-                    street2 = address.street2,
-                    city = address.city,
-                    postalCode = address.postalCode,
-                    country = address.country,
-                    companyName = address.companyName
-                )
+        val shipmentDefinition = ShipmentProviderShipmentDefinition(
+            shipmentId = savedShipment.id!!,
+            ref = shipmentMethod.externalReference,
+            quantity = quantity,
+            weight = weight,
+            address = AddressDefinition(
+                street1 = address.street1,
+                street2 = address.street2,
+                city = address.city,
+                postalCode = address.postalCode,
+                country = address.country,
+                companyName = address.companyName
             )
+        )
+        sendShipmentToExternalProviderWithRetries(webClient, shipmentDefinition)
+    }
+
+    /**
+     * Sends a shipment to an external provider with retries
+     *
+     * @param webClient the web client to use
+     * @param shipmentDefinition the shipment definition to send
+     */
+    private suspend fun sendShipmentToExternalProviderWithRetries(
+        webClient: WebClient,
+        shipmentDefinition: ShipmentProviderShipmentDefinition
+    ) {
+        repeat(experimentConfigService.providerRetries) {
+            try {
+                webClient.post().uri(shipmentProviderConfigurationProperties.endpoint).bodyValue(
+                    shipmentDefinition
+                ).retrieve().toBodilessEntity().awaitSingleOrNull()
+                return
+            } catch (e: Exception) {
+                logger.error("Failed to send shipment to external provider, retry", e)
+                delay(experimentConfigService.providerRetryDelay.toLong())
+            }
+        }
+        webClient.post().uri(shipmentProviderConfigurationProperties.endpoint).bodyValue(
```

Comment:
I do not know how repeat works but wont this try providerRetries + 1 times due to the additional call in line 221?

Reply:
yes
```
    /**
     * The number of retries for the provider
     * Does NOT include the initial request
     */
    var providerRetries: Int = 3
```